src/Symfony/Component/DependencyInjection/Compiler/ReplaceAliasByActualDefinitionPass.php

```diff
@@ -39,98 +38,81 @@ public function process(ContainerBuilder $container)
+        foreach ($container->getDefinitions() as $defId => $definition) {
+            $definition->setArguments($this->updateArgumentReferences($replacements, $defId, $definition->getArguments()));
+            $definition->setMethodCalls($this->updateArgumentReferences($replacements, $defId, $definition->getMethodCalls()));
+            $definition->setProperties($this->updateArgumentReferences($replacements, $defId, $definition->getProperties()));
+            $definition->setFactoryService($this->updateFactoryReferenceId($replacements, $definition->getFactoryService()));
         }
     }
 
     /**
-     * Updates argument references.
+     * Recursively updates references in an array.
      *
-     * @param array  $arguments An array of Arguments
-     * @param string $currentId The alias identifier
-     * @param string $newId     The identifier the alias points to
+     * @param array  $replacements Table of aliases to replace
+     * @param string $defId        Identifier of this definition
+     * @param array  $arguments    Where to replace the aliases
      *
      * @return array
      */
-    private function updateArgumentReferences(array $arguments, $currentId, $newId)
+    private function updateArgumentReferences(array $replacements, $defId, array $arguments)
     {
         foreach ($arguments as $k => $argument) {
             if (is_array($argument)) {
-                $arguments[$k] = $this->updateArgumentReferences($argument, $currentId, $newId);
-            } elseif ($argument instanceof Reference) {
-                if ($currentId === (string) $argument) {
-                    $arguments[$k] = new Reference($newId, $argument->getInvalidBehavior());
-                    $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $this->sourceId, $currentId, $newId));
-                }
+                $arguments[$k] = $this->updateArgumentReferences($replacements, $defId, $argument);
+                continue;
+            }
+            if (!$argument instanceof Reference) {
+                continue;
             }
+            $refId = (string) $argument;
+            if (!isset($replacements[$refId])) {
+                continue;
+            }
+            $newId = $replacements[$refId];
+            $arguments[$k] = new Reference($newId, $argument->getInvalidBehavior());
+            $this->compiler->addLogMessage($this->formatter->formatUpdateReference($this, $defId, $refId, $newId));
         }
 
         return $arguments;
     }
 
-    private function updateFactoryServiceReference($factoryService, $currentId, $newId)
+    /**
+     * Returns the updated reference for the factory service.
+     *
+     * @param array  $replacements Table of aliases to replace
+     * @param string $referenceId  Factory service reference identifier
+     *
+     * @return string
+     */
+    private function updateFactoryReferenceId($replacements, $referenceId)
     {
-        if (null === $factoryService) {
```

Comment:
@SenseException there is nothing to do to define a string as valid. there is work to do to forbid it. And given that we have not written code to forbid it, it means that the empty id is accepted by the code today, and so must continue to work (in case someone had the weird idea to use it).